File: vinokur_code/stretch.cpp
/*
###################################################################################
#
# RIAM-COMPACT
#
# Copyright (C) 2015-2017 Research Institute for Applied Mechanics(RIAM)
#                       / Research Institute for Information Technology(RIIT), Kyushu University.
# All rights reserved.
#
# Copyright (C) 2015-2016 Advanced Institute for Computational Science, RIKEN.
# All rights reserved.
#
###################################################################################
*/

/**
 * @file   riamc_stretch.cpp
 * @brief  Vinokur's stretching function
 * @author RIIT
 */

#include "stretch.h"

#define EPS 1.0e-5
#define ITER_MAX 1000


// @brief 格子分布を反復的に求める
// @param [out] coord 座標値
// @note coordのヌルポインタチェックは事前に
bool Stretch::distribution(REAL_TYPE* coord)
{
  FILE* fp;
  double n1 = (double)(numNodes-1);

  // [0,1]に正規化
  for (int j=0; j<numNodes; j++)
  {
    x [j] = (double)j / n1;
    wk[j] = (double)j / n1;
  }

  // [0,1]で分布を計算するためlengthで正規化
  double sp1 = sp_s / length;
  double sp2 = sp_e / length;

  if ( sp1>1.0 || sp2>1.0) return false;

  double s11 = 1.0 / n1 / sp1;
  double s21 = 1.0 / n1 / sp2;

  stretching(s11, s21);
  int n = numNodes;
  double dx1 = wk[1]   - wk[0];
  double dx2 = wk[n-1] - wk[n-2];

#ifdef _DEBUG
  if ( !(fp=fopen("coord.txt", "w")) )
  {
    printf("\tSorry, can't open 'coord.txt' file.\n");
    exit(1);
  }
#endif

  if ( fabs(sp1-dx1)/sp1<EPS && fabs(sp2-dx2)/sp2<EPS ) goto L10;

  int itr;
  for (itr=0; itr<ITER_MAX; itr++) {
    double ds1 = s11 * 0.1 ;
    double ds2 = s21 * 0.1 ;
    double s12 = s11 + ds1 ;
    double s22 = s21 + ds2 ;

    stretching(s12, s21);
    double dx1_ds1 = (wk[1]  -wk[0]  -dx1) / ds1;
    double dx2_ds1 = (wk[n-1]-wk[n-2]-dx2) / ds1;

    stretching(s11, s22);
    double dx1_ds2 = (wk[1]  -wk[0]  -dx1) / ds2;
    double dx2_ds2 = (wk[n-1]-wk[n-2]-dx2) / ds2;

    double df1 = sp1 - dx1;
    double df2 = sp2 - dx2;

    double d = dx1_ds1 * dx2_ds2 - dx1_ds2 * dx2_ds1;
    ds1 = (df1 * dx2_ds2 - df2 * dx1_ds2) / d;
    ds2 = (df2 * dx1_ds1 - df1 * dx2_ds1) / d;

    s12 = s11 + ds1 ;
    s22 = s21 + ds2 ;

    stretching(s12, s22);
    dx1 = wk[1]   - wk[0];
    dx2 = wk[n-1] - wk[n-2];

#ifdef _DEBUG
    fprintf(fp,"%4d: ", itr);
    for (int i=0; i<numNodes; i++)
    {
      fprintf(fp,"%8.5lf ", wk[i]);
    }
    fprintf(fp,"\n");
#endif

    if ( fabs(sp1-dx1)/sp1<EPS && fabs(sp2-dx2)/sp2<EPS ) break;

    s11 = s12;
    s21 = s22;
  }

  L10:
    wk[0]   = 0.0;
    wk[n-1] = 1.0;


  for (int j=0; j<numNodes; j++)
  {
    coord[j] = (REAL_TYPE) (x_s + wk[j] * length);
  }

#ifdef _DEBUG
  fprintf(fp, "\n\nIteration = %d\n", itr);
  for (int i=0; i<numNodes; i++)
  {
    fprintf(fp,"%3d %8.5lf %8.5f ", i, wk[i], coord[i]);
    if (i==0) {
      fprintf(fp, "\n");
    }
    else {
      fprintf(fp, "%8.5f\n", coord[i]-coord[i-1]);
    }
  }
  fclose(fp);
#endif

  if (itr == ITER_MAX) return false;
  return true;
}


/**
  * @brief 両端がs0, s1の間隔の分布を近似する
  * @param [in]  s0   width of grid at start position [0, 1]
  * @param [in]  s1   width of grid at end position [0, 1]
  */
void Stretch::stretching(const double s0, const double s1)
{
  const double epsilon = 0.001;
  double dz, tanx, tanhx, u;

  double b = sqrt(s0*s1);
  double a = b / s1;

  if ( b < 1.0-epsilon )
  {
    dz = asin(b);
    for (int j=0; j<numNodes; j++)
    {
      tanx = tan( dz * x[j] );
      wk[j] = tanx / (a*sin(dz) + (1.0-a*cos(dz))*tanx);
    }
  }
  else
  {
    if ( b > 1.0+epsilon )
    {
      dz = asinh(b);
      for (int j=0; j<numNodes; j++)
      {
        tanhx = tanh( dz * x[j] ) ;
        wk[j] = tanhx / (a*sinh(dz) + (1.0-a*cosh(dz)) *tanhx);
      }
    }
    else
    {
      for (int j=0; j<numNodes; j++)
      {
        u = x[j] * (1.0+2.0*(b-1.0)*(x[j]-5.0)*(1.0-x[j])); // 5.0 > 0.5?
        wk[j] = u / ( a+(1.0-a)*u);
      }
    }
  }
}

/* deprecated
// asin
double Stretch::f_asin(const double y)
{
  const double a3 = -2.6449341;
  const double a4 = 6.794732;
  const double a5 = -13.205501;
  const double a6 = 11.726095;
  const double b1 = 0.15;
  const double b2 = 0.057321429;
  const double b3 = 0.048774283;
  const double b4 = -0.053337753;
  const double b5 = 0.075845134;
  const double u1 = 0.2693897165164;
  const double pi = 3.14159265358981;
  double ret, yb;

  if ( y < u1 )
  {
    ret = pi*((((((a6*y+a5)*y + a4) *y +a3)*y+1.0)*y-1.0)*y+1.0);
  }
  else
  {
    yb = 1.0 - y;
   ret = sqrt(6.0*yb) * (((((b5*yb+b4)*yb+b3)*yb+b2)*yb+b1)*yb+1.0);
}

  return ret;
}


// asinh
double Stretch::f_asinh(const double y)
{
  const double a1 = 0.15;
  const double a2 = 0.0573214285714;
  const double a3 = -0.024907294878;
  const double a4 = 0.0077424460899;
  const double a5 = -0.0010794122691;
  const double b0 = -0.0204176930892;
  const double b1 = 0.2490272170591;
  const double b2 = 1.9496443322775;
  const double b3 = -2.629454725241;
  const double b4 = 8.5679591096315;
  const double u1 = 2.7829681178603;
  const double u2 = 0.028527431;
  double ret, yb, v, w;

  if ( y < u1 )
  {
    yb = y - 1.0000000;
    ret = sqrt(6.0*yb)*(((((a5*yb+a4)*yb+a3)*yb+a2)*yb+a1)*yb+1.0);
  }
  else
  {
    v = log(y);
    w = 1.0 / y - u2;
    ret = v + log(2.0 * v) *(1.0 + 1.0/v)
        + (((b4 * w +b3)*w+b2)*w+b1)*w+b0;
  }

  return ret;
}
*/

File: Vinokur_Julia/stretch.jl
using Printf
using Plots

const EPS = 1.0e-5
const ITER_MAX = 20

"""
    stretch(x1, x2, sp1, sp2, numNodes; debug=false)

Vinokurのストレッチング関数を用いて格子点分布を計算する

# Arguments
- `x1::Float64`: 開始点座標
- `x2::Float64`: 終了点座標
- `sp1::Float64`: 開始点側の格子点間隔
- `sp2::Float64`: 終了点側の格子点間隔
- `numNodes::Int`: 分割格子点数
- `debug::Bool`: デバッグ出力の有無（デフォルト: false）

# Returns
- `coord::Vector{Float64}`: 格子点座標配列
- `success::Bool`: 収束成功フラグ
"""
function stretch(x1::Float64, x2::Float64, sp1::Float64, sp2::Float64, numNodes::Int; debug::Bool=false)
  length = abs(x2 - x1)
  n1 = Float64(numNodes - 1)
  
  # 初期化：[0,1]に正規化
  x = zeros(Float64, numNodes)
  wk = zeros(Float64, numNodes)
  
  for j in 1:numNodes
    x[j] = Float64(j-1) / n1
    wk[j] = Float64(j-1) / n1
  end
  
  # [0,1]で分布を計算するためlengthで正規化
  sp1_norm = sp1 / length
  sp2_norm = sp2 / length
  
  if sp1_norm > 1.0 || sp2_norm > 1.0
    @warn "格子間隔が全長より大きいです"
    return zeros(Float64, numNodes), false
  end
  
  s11 = 1.0 / n1 / sp1_norm
  s21 = 1.0 / n1 / sp2_norm
  
  stretching!(wk, x, s11, s21)
  dx1 = wk[2] - wk[1]
  dx2 = wk[numNodes] - wk[numNodes-1]
  
  debug_data = []
  
  itr = 0
  if abs(sp1_norm - dx1) / sp1_norm < EPS && abs(sp2_norm - dx2) / sp2_norm < EPS
    @goto convergence
  end
  
  for iter in 1:ITER_MAX
    itr = iter
    ds1 = s11 * 0.1
    ds2 = s21 * 0.1
    s12 = s11 + ds1
    s22 = s21 + ds2
    
    # 数値微分による勾配計算
    stretching!(wk, x, s12, s21)
    dx1_ds1 = (wk[2] - wk[1] - dx1) / ds1
    dx2_ds1 = (wk[numNodes] - wk[numNodes-1] - dx2) / ds1
    
    stretching!(wk, x, s11, s22)
    dx1_ds2 = (wk[2] - wk[1] - dx1) / ds2
    dx2_ds2 = (wk[numNodes] - wk[numNodes-1] - dx2) / ds2
    
    df1 = sp1_norm - dx1
    df2 = sp2_norm - dx2
    
    # ニュートン法の更新
    d = dx1_ds1 * dx2_ds2 - dx1_ds2 * dx2_ds1
    if abs(d) < 1e-15
      @warn "ヤコビアンが特異です"
      return zeros(Float64, numNodes), false
    end
    
    ds1 = (df1 * dx2_ds2 - df2 * dx1_ds2) / d
    ds2 = (df2 * dx1_ds1 - df1 * dx2_ds1) / d
    
    s12 = s11 + ds1
    s22 = s21 + ds2
    
    stretching!(wk, x, s12, s22)
    dx1 = wk[2] - wk[1]
    dx2 = wk[numNodes] - wk[numNodes-1]
    
    if debug
      push!(debug_data, copy(wk))
      @printf("Iter %2d: dx1=%8.5f dx2=%8.5f error1=%8.5e error2=%8.5e\n", 
              iter, dx1, dx2, abs(sp1_norm-dx1)/sp1_norm, abs(sp2_norm-dx2)/sp2_norm)
    end
    
    if abs(sp1_norm - dx1) / sp1_norm < EPS && abs(sp2_norm - dx2) / sp2_norm < EPS
      break
    end
    
    s11 = s12
    s21 = s22
  end
  
  @label convergence
  # 境界値を正確に設定
  wk[1] = 0.0
  wk[numNodes] = 1.0
  
  # 実座標系にスケールバック
  coord = zeros(Float64, numNodes)
  for j in 1:numNodes
    coord[j] = x1 + wk[j] * length
  end
  
  if debug
    println("\n=== 収束結果 ===")
    @printf("反復回数: %d\n", itr)
    println("格子点座標:")
    for i in 1:numNodes
      if i == 1
        @printf("%3d %8.5f %8.5f\n", i-1, wk[i], coord[i])
      else
        @printf("%3d %8.5f %8.5f %8.5f\n", i-1, wk[i], coord[i], coord[i]-coord[i-1])
      end
    end
  end
  
  success = (itr < ITER_MAX)
  return coord, success
end

"""
両端がs0, s1の間隔の分布を近似する内部関数
"""
function stretching!(wk::Vector{Float64}, x::Vector{Float64}, s0::Float64, s1::Float64)
  epsilon = 0.001
  numNodes = length(x)
  
  b = sqrt(s0 * s1)
  a = b / s1
  
  if b < 1.0 - epsilon
    # asin分枝
    dz = asin(b)
    for j in 1:numNodes
      tanx = tan(dz * x[j])
      wk[j] = tanx / (a * sin(dz) + (1.0 - a * cos(dz)) * tanx)
    end
  elseif b > 1.0 + epsilon
    # asinh分枝
    dz = asinh(b)
    for j in 1:numNodes
      tanhx = tanh(dz * x[j])
      wk[j] = tanhx / (a * sinh(dz) + (1.0 - a * cosh(dz)) * tanhx)
    end
  else
    # 線形近似分枝
    for j in 1:numNodes
      u = x[j] * (1.0 + 2.0 * (b - 1.0) * (x[j] - 0.5) * (1.0 - x[j]))
      wk[j] = u / (a + (1.0 - a) * u)
    end
  end
end

"""
格子分布を可視化する関数
"""
function plot_grid_distribution(coord::Vector{Float64}, title_str::String="Grid Distribution")
  numNodes = length(coord)
  
  # 格子点間隔を計算
  spacings = zeros(Float64, numNodes-1)
  for i in 1:numNodes-1
    spacings[i] = coord[i+1] - coord[i]
  end
  
  # プロット作成
  p = plot(layout=(2,1), size=(800, 600))
  
  # 上段：格子点分布
  scatter!(p[1], coord, ones(numNodes), 
           xlabel="Position", ylabel="", 
           title=title_str, 
           legend=false, 
           markersize=3, 
           ylims=(0.5, 1.5))
  
  # 下段：格子点間隔
  plot!(p[2], 1:numNodes-1, spacings, 
        xlabel="Grid Index", ylabel="Grid Spacing", 
        title="Grid Spacing Distribution", 
        legend=false, 
        linewidth=2, 
        marker=:circle, 
        markersize=3)
  
  return p
end